### PR TITLE
Projection mapping

### PR DESCRIPTION
**Pull request by haggi**
Make projection mapping work again
Update copyright info in osl files
Reimplement automatic tx texture file conversion (if enabled in render globals)

### CODE REVIEW

File: src/utilities/oslutils.cpp [hunk 1]
```diff
@@ -505,6 +505,12 @@ void OSLUtilClass::defineOSLParameter(ShaderAttribute& sa, MFnDependencyNode& de
     if (sa.type == "matrix")
     {
         MMatrix value = getMatrix(plug);
+        boost::shared_ptr<RenderGlobals> renderGlobals = getWorldPtr()->mRenderGlobals;
+        MMatrix scale;
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: You can simplify and also avoid a matrix multiplication if `renderGlobals` evaluates to false:

    if (renderGlobals)
        value *= renderGlobals->globalConversionMatrix;

But can `renderGlobals` really evaluate to false?

File: src/utilities/oslutils.cpp [hunk 2]
```diff
@@ -530,6 +536,16 @@ void OSLUtilClass::addConnectionToConnectionArray(ConnectionArray& ca, MString s
     ca.push_back(c);
 }
 
+// Projection nodes in Maya and OLS need a special handling. Other shading systems allow the manipulation of the global u and v or s and t values
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: *OSL*

File: src/utilities/oslutils.cpp [hunk 3]
```diff
@@ -530,6 +536,16 @@ void OSLUtilClass::addConnectionToConnectionArray(ConnectionArray& ca, MString s
...
+// before they access a texture. This is not possible in OSL what means we have to feed the correct uv values into a 2d texture node. In the case of a
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: *which means*

File: src/utilities/oslutils.cpp [hunk 4]
```diff
@@ -530,6 +536,16 @@ void OSLUtilClass::addConnectionToConnectionArray(ConnectionArray& ca, MString s
...
 
+// Projection nodes in Maya and OLS need a special handling. Other shading systems allow the manipulation of the global u and v or s and t values
+// before they access a texture. This is not possible in OSL what means we have to feed the correct uv values into a 2d texture node. In the case of a 
+// projection node, we first have to find out which 2d nodes are connected to our node directly or indirectly. Then we create a projection helper node
+// which calculates the uv data and plug it into the 2d node. Later we use another instance of the same projection node to calculate color balance and 
+// other color specific elements.
+// 
+// placementMatrixNode -> projection_util -> 2dnode -> projectionNode -> output
+//
+// This method as a small drawback. It is not possible to use the same 2d texture node for different projections.
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: *has*

File: src/utilities/oslutils.cpp [hunk 5]
```diff
@@ -1116,18 +1133,15 @@ void OSLUtilClass::initOSLUtil()
...
     {
         if (projNode == projectionConnectNodes[i])
         {
-            MString sourceNode = (getObjectName(projectionNodes[i]) + "_ProjUtil");
-            MString sourceAttr = "outUVCoord";
-            MString destNode = pn.name();
-            MString destAttr = "uvCoord";
-            ca.push_back(Connection(sourceNode, sourceAttr, destNode, destAttr));
-            connectOSLShaders(ca);
+            const MString sourceNode = (getObjectName(projectionNodes[i]) + "_ProjUtil");
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: Remove redundant parentheses.

File: src/utilities/oslutils.cpp [hunk 6]
```diff
@@ -1169,9 +1183,9 @@ namespace
         if (param.type == OSL::TypeDesc::TypeMatrix)
         {
             SimpleMatrix &v = boost::get<SimpleMatrix>(param.value);
-            result = MString("matrix ") + v.f[0][0] + " " + v.f[0][1] + " " + v.f[0][2] + " " + v.f[0][3] +
-                v.f[1][0] + " " + v.f[1][1] + " " + v.f[1][2] + " " + v.f[1][3] +
-                v.f[2][0] + " " + v.f[2][1] + " " + v.f[2][2] + " " + v.f[2][3] +
+            result = MString("matrix ") + v.f[0][0] + " " + v.f[0][1] + " " + v.f[0][2] + " " + v.f[0][3] +" " +
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: What about using the new `format()` function here?

File: src/appleseedrenderer.cpp [hunk 7]
```diff
@@ -885,6 +884,9 @@ void AppleseedRenderer::doInteractiveUpdate()
     for (iaIt = interactiveUpdateList.begin(); iaIt != interactiveUpdateList.end(); iaIt++)
     {
         InteractiveElement *iElement = *iaIt;
+        if (iElement == 0)
```
Comment (haggi) [inline src/appleseedrenderer.cpp]: If the render region changes in IPR, the renderer waits for an update in the interactiveUpdateList to start again. In this case we have no interactive element, we only want to trigger the rendering.

File: src/utilities/oslutils.cpp (hunk 1 above)
Comment (haggi) [inline src/utilities/oslutils.cpp]: Yes, render globals are not defined, or may not be defined if a swatch rendering or Maya2016 material preview is created.

File: src/appleseedrenderer.cpp (hunk 7 above)
Comment (dictoon) [inline src/appleseedrenderer.cpp]: But wouldn't `interactiveUpdateList` be empty then, rather?

Comment (haggi) [inline src/appleseedrenderer.cpp]: That's the reason why I put an null element into the list. This triggers a rendering, but has no other use.

Comment (dictoon) [inline src/appleseedrenderer.cpp]: Ah, I see. Maybe we can a more elegant method, but in the meantime this explanation would stand very well above the `if` :)

File: src/utilities/oslutils.cpp [hunk 8]
```diff
@@ -1183,10 +1182,14 @@ namespace
         if (param.type == OSL::TypeDesc::TypeMatrix)
         {
             SimpleMatrix &v = boost::get<SimpleMatrix>(param.value);
-            result = MString("matrix ") + v.f[0][0] + " " + v.f[0][1] + " " + v.f[0][2] + " " + v.f[0][3] +" " +
-                v.f[1][0] + " " + v.f[1][1] + " " + v.f[1][2] + " " + v.f[1][3] + " " +
-                v.f[2][0] + " " + v.f[2][1] + " " + v.f[2][2] + " " + v.f[2][3] + " " +
-                v.f[3][0] + " " + v.f[3][1] + " " + v.f[3][2] + " " + v.f[3][3];
+            result = MString("matrix");
+            for (unsigned int i = 0; i < 4; i++)
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: The point here was to use a single `format()` call, but I didn't realize we only supported 4 arguments, not 16 :) So I guess this is good enough for now.

**Comment by dictoon** [comment]
Thank you, great code!

**Pull request merged**